ICRC-97: Non-Fungible Token (NFT) Metadata Standard

sea-snake opened the pull request:
Standardize approach 2 at https://forum.dfinity.org/t/approaches-to-icrc-7-nft-user-facing-metadata/42884

sea-snake committed:
Create ICRC-97.md

sea-snake committed:
Update ICRC-97.md

dietersommer commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,95 @@
+
+The standard is designed to allow for multiple assets with different data types with various purposes. Where additional
+asset purposes and attribute display types can be standardized in ICRC-97 extensions.
+
```
Should we add the main uses of the standard here as motivation, e.g., that it allows marketplaces to properly show the NFTs?

sea-snake replied:
Good point, added!

dietersommer commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,95 @@
+pairs.
+
+ICRC-7 does not specify the representation of token metadata any further than that it is represented in a generic manner
+as a vector of (text, Value)-pairs. The ICRC-97 standard extends the ICRC-7 standard with a token metadata
```
We could mention that the (text, Value)-pair rep is recursive.

sea-snake replied:
Added a sentence to mention that `Value` is recursive as defined in the ICRC-3 standard.

dietersommer commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,95 @@
+The standard is designed to allow for multiple assets with different data types with various purposes. Where additional
+asset purposes and attribute display types can be standardized in ICRC-97 extensions.
+
+## Extending the ICRC-7 Standard with Token Metadata
```
This would also apply to future NFT standards such as ICRC-8, right? In this case, we should at least mention this, maybe also mention ICRC-8 as potential client (in future work section).

sea-snake replied:
Rephrased it to indicate that ICRC-97 is not solely for ICRC-7.

dietersommer commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,95 @@
+
+### Metadata properties
+
+Below table list all (optional) metadata properties defined in this standard. Each property is defined in both ICRC-3
```
"list" -> "lists"

dietersommer commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,95 @@
+### Metadata properties
+
+Below table list all (optional) metadata properties defined in this standard. Each property is defined in both ICRC-3
+and JSON type to support both on-chain `Value` and off-chain JSON metadata respectively.
```
"metadata respectively." -> "metadata, respectively."

dietersommer commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,95 @@
+| icrc97:image   | Original size image that is shown on e.g. item details page. Additional optional `width` and `height` properties define the dimensions in number of pixels. |
+| icrc97:preview | Small image meant as preview within e.g. a list of items. Additional optional `width` and `height` properties define the dimensions in number of pixels.    |
+
+The following table list all (optional) attribute properties.
```
"list" -> "lists"

sea-snake committed:
Update ICRC-97.md

Indicate that the minimum list of asset purposes and display types are defined.

sea-snake committed:
Update ICRC-97.md

Rephrase purpose of ICRC-97.

sea-snake committed:
Update ICRC-97.md

Mention Value is recursive.

sea-snake committed:
Update ICRC-97.md

Fix spelling

sea-snake committed:
Update ICRC-97.md

Fix spelling

sea-snake committed:
Update ICRC-97.md

Improve phrasing.

sea-snake committed:
Update ICRC-97.md

Mention future work.

marc0olo commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+Below list of purpose values are part of the ICRC-97 standard, this list could be extended by other standards.
+Purpose values can define additional asset properties.
+
+| Purpose Value  | Description                                                                                                                                                 |
```
can you provide an example on how this would be defined?

marc0olo commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+
+| Purpose Value  | Description                                                                                                                                                 |
+|----------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| icrc97:image   | Original size image that is shown on e.g. item details page. Additional optional `width` and `height` properties define the dimensions in number of pixels. |
```
can you provide an example on how this would be defined?

sea-snake replied:
I'll add examples to the spec

marc0olo commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+| icrc97:property         | Shows attribute as property with e.g. rarity, this is the default display type.                                                                               |
+| icrc97:date             | Shows attribute as date, expects epoch timestamp number in milliseconds.                                                                                      |
+| icrc97:time             | Shows attribute as date with time, expects epoch timestamp number in milliseconds.                                                                            |
+| icrc97:rank             | Shows attribute as progress rank e.g. 4 of 10, expects number and additional `max_value` property with a number.                                              |
```
can you provide an example on how 4 of 10 would be defined?

sea-snake replied:
I'll add examples to the spec

marc0olo commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+| name              | Yes      | variant { Text }                        | string    | Plain text.                                             |
+| description       | Yes      | variant { Text }                        | string    | Markdown.                                               |
+| assets            | Yes      | variant { Array = vec variant { Map } } | object[]  | List of assets ordered by priority descending.          |
+| attributes        | Yes      | variant { Array = vec variant { Map } } | object[]  | List of attributes ordered by priority descending.      |
```
are the attributes expected to be immutable? otherwise we might want to distinguish between mutable and immutable attributes

marc0olo replied:
the term "immutable" is of course tricky, given the nature of canisters living on ICP.

did you generally discuss to track the history of metadata changes so far? 🤔

sea-snake replied:
Good points, there's the [update schema in ICRC-7](https://github.com/dfinity/ICRC/blob/icrc7-wg-draft/ICRCs/ICRC-7/ICRC-7.md#icrc7_update-block-schema) for the purpose of tracking metadata updates. I'll mention how to make the metadata immutable, basically by not returning above block type in [ICRC-3 supported block types](icrc3_supported_block_types). I'll also mention on how ICRC-97 metadata updates should be encoded as blobs in the update schema in case the metadata is not immutable.

marc0olo replied:
I think I need to read more about ICRC-3 🤔 

note: the link is not working.

sea-snake replied:
> note: the link is not working.

Fixed

letmejustputthishere commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+The following metadata property MUST be defined in the root of the token metadata if we want to return on-chain
+metadata.
+
+| Root Metadata Property | ICRC-3 Type     | Description                            |
+|------------------------|-----------------|----------------------------------------|
+| icrc97:metadata        | variant { Map } | Contains on-chain metadata properties. |
+
+Alternatively the following metadata property MUST be defined in the root of the token metadata if we want to return
+external
+off-chain metadata instead of on-chain metadata.
+
+| Root Metadata Property   | ICRC-3 Type     | Description                                       |
+|--------------------------|-----------------|---------------------------------------------------|
+| icrc97:external_metadata | variant { Map } | Contains below external JSON metadata properties. |
```
this does not prevent defining both, right?

sea-snake replied:
The metadata ICRC-3 structure doesn't prevent any dev from not following the spec, but you're right, according to the ICRC-97 spec only of these should be defined. I'll clarify this in the spec.

sea-snake commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+| external_url      | Yes      | variant { Text }                        | string    | URL that allows the user to view the item on your site. |
+| name              | Yes      | variant { Text }                        | string    | Plain text.                                             |
+| description       | Yes      | variant { Text }                        | string    | Markdown.                                               |
+| assets            | Yes      | variant { Array = vec variant { Map } } | object[]  | List of assets ordered by priority descending.          |
```
> List of assets ordered by priority descending.

The order of the list of assets matters, dapps are expected to prioritize assets higher in the list.

marc0olo replied:
so as a consumer (e.g. explorer or marketplace) I would use the first occurrence of an asset with type image (or preview if existing), right? assuming that the NFT has an image attached.

if we can have more than one asset with purpose `irc97:preview`, I'd expect that `irc97:preview` differs from its corresponding `irc97:image`. otherwise the consumer could just resize the `irc97:image` to the needs.

sea-snake replied:
> I would use the first occurrence of an asset with type image (or preview if existing), right?

Yes, I'll clarify this in the spec a bit further.

> otherwise the consumer could just resize the irc97:image to the needs

I'll also clarify in the spec that `irc97:image` should be used if `irc97:preview` is missing.

sea-snake commented on ICRCs/ICRC-97/ICRC-97.md:
```diff
@@ -0,0 +1,101 @@
+| External Metadata Property | Optional | ICRC-3 Type      | Description                                                                    |
+|----------------------------|----------|------------------|--------------------------------------------------------------------------------|
+| url                        | No       | variant { Text } | URL that returns the metadata in JSON format (protocol is not limited to HTTP) |
+| sha256_hash                | Yes      | variant { Blob } | SHA-256 hash of HTTP response body bytes returned from above url.              |
```
Yes that's what I meant with `SHA-256 hash of HTTP response body bytes returned from above url.`. 

I'll clarify this further, keep in mind the hash is optional, so a dev can opt to not have it. 

Overall, immutability on the IC is only truly something one could trust if the contract code is public, build are reproducible and the canister doesn't have controllers defined (or blackholed). As I mentioned earlier, some method(s) could indicate that things are immutable, but unless above is true, you can't fully trust it.

sea-snake committed:
Refactor standard to align with JSON approach.

sea-snake committed:
Refactor standard to align with JSON approach.

sea-snake committed:
Refactor standard to align with JSON approach.

sea-snake committed:
Remove ICRC-91 

Remove ICRC-91 since it's not finalized yet and doesn't need to be in this standard.
